ambry-router/src/main/java/com/github/ambry/router/PutOperation.java

```diff
@@ -305,6 +314,9 @@ private PutOperation(RouterConfig routerConfig, NonBlockingRouterMetrics routerM
     isEncryptionEnabled = passedInBlobProperties.isEncrypted();
     restRequest = options.getRestRequest();
     loggingContext = makeLoggingContext();
+
+    // TODO - use dependency injection when available.
+    compressionService = new CompressionService(routerConfig.getCompressionConfig(), routerMetrics.compressionMetrics);
```

Comment:
we should create one CompressionService object and pass it PutOperation. CompressionService's constructor has some logics that will be annoy to see them run every single time when creating a PutOperation, like logging an error about default Compression, and setting PutRequest.Version.

Reply:
CompressionService should be a singleton.  If we create it in operation manager and pass it to every PutOperation, that implies it's a singleton.  In my opinion, the best place for singleton is Dependency Injection which we don't have yet, but not through parameter passing.  In this case, passing a singleton to PutOperation constructor does not add "useful value" to the logic.  Changing many unit tests to match the new constructor also has no value.  In here, I define "value" as additional feature or improvements.  It is not a good idea to write/change code that has no value just to meet the compiler syntax requirement.  It will just occupy developer and reviewers time in the long run.  When we have changes the has wide impact, it is a signal that the coding structure might need improvement.

I will make the suggested change here until we have Dependency Injection. Together, let's continue to make improvements.